### ISSUE TITLE
Migrate parameters to torch.nn.Modules

### ISSUE THREAD

**Issue by cpehle** [norse/norse#143]
Currently neuron parameters such as the synaptic time constant and neuron time constants are not trained. Recent experimental work suggests that this enables better performance on a number of benchmark tasks. We should investigate trainable neuron parameters in Norse as well. This shouldn't really require code changes, just be evaluated as an option.

This excellent talk by Sander Bohte explains the role of trainable adaptation well:
https://www.youtube.com/watch?v=5SrEycIbfRE

**Comment by Huizerd** [comment]
Maybe something like evolution or some other gradient-free optimization? Like https://github.com/facebookresearch/nevergrad

**Comment by cpehle** [comment]
That is certainly an option although I think it would be possible toset requires_grad=True in say the time constants of the LIFParameters https://github.com/norse/norse/blob/a2c15bbe59b01745453b579c2ec57135674d2093/norse/torch/functional/lif.py#L26-L27,
at least that is what I understand the talk by Sander Bohte talks about. Would be interesting to try out on a harder task like speech commands or some other time series task, where he reports success with adaptive LIF neurons and trainable time constants.

**Comment by Jegp** [comment]
This paper might be a source of inspiration: https://www.researchgate.net/publication/345981609_Incorporating_Learnable_Membrane_Time_Constant_to_Enhance_Learning_of_Spiking_Neural_Networks

**Comment by dsalaj** [comment]
Another recent paper focused on jointly training neuron parameters and weights: https://arxiv.org/pdf/2010.12691.pdf

**Comment by Jegp** [comment]
I updated the title to reflect the idea of @Huizerd that neuron parameters could be a module instead of a NamedTuple, which would make it much easier to train neuron parameters and do all sorts of gradient gymnastics.

Theoretically, there's nothing preventing us from learning parameters now, as illustrated in the [Parameter Learning Notebook](https://github.com/norse/norse-hbp-workshop).

**Comment by Jegp** [comment]
We discussed the parameter learning and all the building blocks actually exist to make it work currently.

We agreed to improve our tutorial and add documentation to explain how it can be done, instead of trying to be "clever" and add some functionality that might miss the target.

I would suggest we close the issue for now. If there are any good ideas on how to improve the current situation, we are all ears! :ear:  :-)
